proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -0,0 +1,233 @@
+specified resource class. 
+
+| Resource Class | Register Type | Diagnostic |
+|-|-|-|
+| Sampler | s | "Object type '%0' with resource class "Sampler" expects register type 's', but register type '%1' was given." |
+| SRV | t | "Object type '%0' with resource class "SRV" expects register type 't', but register type '%1' was given." |
+| UAV | u | "Object type '%0' with resource class "UAV" expects register type 'u', but register type '%1' was given." |
+| CBuffer | b |  "Object type '%0' with resource class "CBuffer" expects register type 'b', but register type '%1' was given." |
+
+If the given candidate resource is a user-defined type (UDT), then further
+analysis is necessary. The first step is to gather all registers that
+are being bound to this declaration, and collect the register types
+that are being specified. The UDT must have at least one valid resource
+that can be bound to the provided register type(s). If not, an error must be
+emitted stating that "No resource contained in struct '%0' can be bound
+to register type '%1'". There are no issues if a UDT has more resources
+than there are register binding statements, the resources will be bound to
+the next available space automatically, and so compilation can succeed.
+Additionally, a singular resource can be bound multiple times to different
+spaces, so multiple register types are allowed for a single resource.
+Below are some examples of different UDT's and the diagnostics that
+would be emitted when they are bound:
+
+```
+struct Eg1 {
+  float f;
+  Buffer<float> Buf;
+  RWBuffer<float> RWBuf;
+  };
+Eg1 e1 : register(t0) : register(s0); 
+// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to s0
+
+struct Eg2 {
+  float f;
+  Buffer<float> Buf;
+  RWBuffer<float> RWBuf;
+  RWBuffer<float> RWBuf2;
+  };
+Eg2 e2 : register(t0) : register(s0); 
+// Valid: f is skipped, Buf is bound to t0, RWBuf is bound to s0. RWBuf2 gets assigned to s1 even though there is no explicit binding for s1.
+
+struct Eg3 {
+  float f;
+  Buffer<float> Buf;
+  }; 
+Eg3 e3 : register(t0) : register(s0);
+// Valid: Buf gets bound to t0. Buf will also be bound to s0.
+
+struct Eg4 {
+  struct Bar {
+    RWBuffer<int> a;
+    };
+    Bar b;
+};
+Eg4 e4 : register(t0) 
+// Valid: Bar, the struct within Eg4, has a valid resource that can be bound to t0. 
+
+struct Eg5 {
+  float f;
+}; 
+Eg5 e5 : register(t0) 
+// DefaultError warning: "UDT resource 'Eg5' does not contain an applicable resource type for register type 't'"
+
+struct Eg6 {
+  struct Bar {
+      float f;
+    }
+    Bar b;
+};
+Eg6 e6 : register(t0) 
+// DefaultError warning: "UDT resource 'Eg6' does not contain an applicable resource type for register type 't'"|
+
+```
+
+Finally, if the candidate type is not a valid resource type or not a UDT, the
+final case will be entered. Types that are or contain a resource are known as
+"intangible". In this case, we are dealing with types that cannot be intangible.

```

Comment:
I do think `intangible` is the proper word to use here. Do you have an alternate phrase to use to describe types that don't have the resource attribute or do not contain any types that have the resource attribute? Ideally, the spec can refer to this concept with a single word, rather than needing to repeat that whole phrase several times.

Reply:
You're right that we could use some compact language to refer to these types.  These basic types are "tangible" (or not "intangible), but that also applies to UDTs that don't contain intangible types.  I'd like to think of these as "basic" types in HLSL, but might conflict with Clang, since "basic" types in Clang probably don't include our vector and matrix types.